src/shared/components/home/login.tsx

```diff
@@ -20,7 +33,69 @@ interface State {
   siteRes: GetSiteResponse;
 }
 
-export class Login extends Component<any, State> {
+async function handleLoginSubmit(i: Login, event: any) {
+  event.preventDefault();
+  const { password, totp_2fa_token, username_or_email } = i.state.form;
+
+  if (username_or_email && password) {
+    i.setState({ loginRes: { state: "loading" } });
+
+    const loginRes = await HttpService.client.login({
+      username_or_email,
+      password,
+      totp_2fa_token,
+    });
+    switch (loginRes.state) {
+      case "failed": {
+        if (loginRes.msg === "missing_totp_token") {
+          i.setState({ showTotp: true });
+          toast(I18NextService.i18n.t("enter_two_factor_code"), "info");
+        }
+
+        i.setState({ loginRes: { state: "failed", msg: loginRes.msg } });
+        break;
+      }
+
+      case "success": {
+        UserService.Instance.login({
+          res: loginRes.data,
+        });
+        const site = await HttpService.client.getSite({
+          auth: myAuth(),
+        });
+
+        if (site.state === "success") {
+          UserService.Instance.myUserInfo = site.data.my_user;
+        }
+
+        const { prev } = getLoginQueryParams();
+
+        prev
+          ? i.props.history.replace(prev)
+          : i.props.history.action === "PUSH"
+          ? i.props.history.back()
+          : i.props.history.replace("/");
+
+        break;
+      }
+    }
+  }
+}
+
+function handleLoginUsernameChange(i: Login, event: any) {
+  i.state.form.username_or_email = event.target.value.trim();
+  i.setState(i.state);
```

Comment:
Use the single line setState we do elsewhere.

Also it seems like you added these functions, but didn't remove the class ones.